### PR TITLE
matrix property + example

### LINKED ISSUE #189: Extend pipeline reference with matrix feature
For example, this would create `n` concurrent jobs with the same commands but different env vars.
```yaml
      jobs:
        - name: matrix job
          commands:
            - echo $ELIXIR
            - echo $ERLANG
          matrix:
            - env_var: ELIXIR
              values: ["1.3", "1.4"]
            - env_var: ERLANG
              values: ["19", "20", "21"]
```

There will be 6 jobs created because there are 6 elements in Cartesian product of env vars in matrix property:

- `ELIXIR=1.3` and `ERLANG=19`
- `ELIXIR=1.3` and `ERLANG=20`
- `ELIXIR=1.3` and `ERLANG=21`
- `ELIXIR=1.4` and `ERLANG=19`
- `ELIXIR=1.4` and `ERLANG=20`
- `ELIXIR=1.4` and `ERLANG=21`

### PR DESCRIPTION
**Pull request by mactsouk**
Closes #189

### CODE REVIEW

File: pipeline-yaml_5b5744d40428631d7a8940a9.md [hunk 1]
```diff
@@ -615,6 +614,53 @@ were also defined on the `task` level.
 	          - echo $VAR_1
 	          - echo $APP_ENV
 
+### matrix
+
+The `martix` property allows you to define one or more environment variables
+with multiple values. After that, it creates multiple jobs by calculating the
```
Comment (markoa) [inline pipeline-yaml_5b5744d40428631d7a8940a9.md]: multiple parallel jobs

File: pipeline-yaml_5b5744d40428631d7a8940a9.md [hunk 2]
```diff
@@ -615,6 +614,53 @@ were also defined on the `task` level.
...
+Cartesian product of all the sets of user defined values.
+
+So, the final outcome of the `matrix` property is concurrent jobs with the
```
Comment (markoa) [inline pipeline-yaml_5b5744d40428631d7a8940a9.md]: is concurrent jobs => are parallel jobs

File: pipeline-yaml_5b5744d40428631d7a8940a9.md [hunk 3]
```diff
@@ -615,6 +614,53 @@ were also defined on the `task` level.
...
+commands that are defined in the respective `jobs` property but different
```
Comment (markoa) [inline pipeline-yaml_5b5744d40428631d7a8940a9.md]: "but different environment variables that are defined in the `matrix` property."
=>
joined with all possible combinations of environment variables defined in the `matrix` property.

File: pipeline-yaml_5b5744d40428631d7a8940a9.md [hunk 4]
```diff
@@ -615,6 +614,53 @@ were also defined on the `task` level.
 	          - echo $VAR_1
 	          - echo $APP_ENV
 
+### matrix
+
+The `martix` property allows you to define one or more environment variables
+with multiple values. After that, it creates multiple jobs by calculating the
+Cartesian product of all the sets of user defined values.
```
Comment (markoa) [inline pipeline-yaml_5b5744d40428631d7a8940a9.md]: simplify to "Cartesian product of all sets of values"

(a user is reading this so it doesn't make sense to read user-defined in that perspective)

File: pipeline-yaml_5b5744d40428631d7a8940a9.md [hunk 5]
```diff
@@ -615,6 +614,53 @@ were also defined on the `task` level.
...
+	          commands:
+	            - echo $ELIXIR
+	            - echo $ERLANG
+	          matrix:
+	            - env_var: ELIXIR
+	              values: ["1.3", "1.4"]
+	            - env_var: ERLANG
+	              values: ["19", "20", "21"]
+
+This will create six jobs because there are 6 elements in the Cartesian product
```
Comment (markoa) [inline pipeline-yaml_5b5744d40428631d7a8940a9.md]: Let's not repeat the mathematical term nobody remembers. :)

We can simplify both paragraphs to one sentence: _In this example the `matrix` expands to 6 parallel jobs as there are 2 * 3 = 6 combinations of provided environment variables:_ (followed by the list)

File: pipeline-yaml_5b5744d40428631d7a8940a9.md (hunk 1 above)
Comment (mactsouk) [reply pipeline-yaml_5b5744d40428631d7a8940a9.md]: Done.

File: pipeline-yaml_5b5744d40428631d7a8940a9.md (hunk 2 above)
Comment (mactsouk) [reply pipeline-yaml_5b5744d40428631d7a8940a9.md]: Done.

File: pipeline-yaml_5b5744d40428631d7a8940a9.md (hunk 3 above)
Comment (mactsouk) [reply pipeline-yaml_5b5744d40428631d7a8940a9.md]: Done.

File: pipeline-yaml_5b5744d40428631d7a8940a9.md (hunk 4 above)
Comment (mactsouk) [reply pipeline-yaml_5b5744d40428631d7a8940a9.md]: Good point.

File: pipeline-yaml_5b5744d40428631d7a8940a9.md (hunk 5 above)
Comment (mactsouk) [reply pipeline-yaml_5b5744d40428631d7a8940a9.md]: Good point.

File: pipeline-yaml_5b5744d40428631d7a8940a9.md [hunk 6]
```diff
@@ -615,6 +614,51 @@ were also defined on the `task` level.
 	          - echo $VAR_1
 	          - echo $APP_ENV
 
+### matrix
+
+The `martix` property allows you to define one or more environment variables
+with multiple values. After that, it creates multiple parallel jobs by
+calculating the Cartesian product of all sets of values.
```
Comment (predrag-rakic) [inline pipeline-yaml_5b5744d40428631d7a8940a9.md]: The `martix` property allows you to define one or more *environment variable sets*
(env var set contains multiple values for that env var).
In such setup `n` parallel jobs are created, where `n` equals cardinality of Cartesian product of all env var sets.

File: pipeline-yaml_5b5744d40428631d7a8940a9.md [hunk 7]
```diff
@@ -615,6 +614,51 @@ were also defined on the `task` level.
...
+
+So, the final outcome of the `matrix` property is parallel jobs with the
+commands that are defined in the respective `jobs` property joined with all
+possible combinations of the environment variables defined in the `matrix`
+property.
```
Comment (predrag-rakic) [inline pipeline-yaml_5b5744d40428631d7a8940a9.md]: So, the final outcome of the `matrix` property is creation of multiple parallel jobs with the
same commands, defined in the respective `commands` property.
Each job is assigned env vars from corresponding element of Cartesian product.

File: pipeline-yaml_5b5744d40428631d7a8940a9.md [hunk 8]
```diff
@@ -615,6 +614,51 @@ were also defined on the `task` level.
...
+	            - echo $ELIXIR
+	            - echo $ERLANG
+	          matrix:
+	            - env_var: ELIXIR
+	              values: ["1.3", "1.4"]
+	            - env_var: ERLANG
+	              values: ["19", "20", "21"]
+
+In this example the values of the `matrix` property expand to 6 parallel jobs
+as there are 2 * 3 = 6 combinations of the provided environment variables:
```
Comment (predrag-rakic) [inline pipeline-yaml_5b5744d40428631d7a8940a9.md]: In this example, job specification named `Elixir + Erlang matrix` expands to 6 parallel jobs
as there are 2 * 3 = 6 combinations of the provided environment variables:

File: pipeline-yaml_5b5744d40428631d7a8940a9.md (hunk 6 above)
Comment (mactsouk) [reply pipeline-yaml_5b5744d40428631d7a8940a9.md]: Done.

File: pipeline-yaml_5b5744d40428631d7a8940a9.md (hunk 7 above)
Comment (mactsouk) [reply pipeline-yaml_5b5744d40428631d7a8940a9.md]: Done.

File: pipeline-yaml_5b5744d40428631d7a8940a9.md (hunk 8 above)
Comment (mactsouk) [reply pipeline-yaml_5b5744d40428631d7a8940a9.md]: Done.

**Pull request merged**